// @effect-atom/atom-react/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Atom: true,
  Registry: true,
  Result: true,
  AtomRef: true,
  AtomHttpApi: true,
  AtomRpc: true,
  Hydration: true
};
exports.Result = exports.Registry = exports.Hydration = exports.AtomRpc = exports.AtomRef = exports.AtomHttpApi = exports.Atom = void 0;
var _Atom = /*#__PURE__*/_interopRequireWildcard(/*#__PURE__*/require("@effect-atom/atom/Atom"));
exports.Atom = _Atom;
var _Registry = /*#__PURE__*/_interopRequireWildcard(/*#__PURE__*/require("@effect-atom/atom/Registry"));
exports.Registry = _Registry;
var _Result = /*#__PURE__*/_interopRequireWildcard(/*#__PURE__*/require("@effect-atom/atom/Result"));
exports.Result = _Result;
var _AtomRef = /*#__PURE__*/_interopRequireWildcard(/*#__PURE__*/require("@effect-atom/atom/AtomRef"));
exports.AtomRef = _AtomRef;
var _AtomHttpApi = /*#__PURE__*/_interopRequireWildcard(/*#__PURE__*/require("@effect-atom/atom/AtomHttpApi"));
exports.AtomHttpApi = _AtomHttpApi;
var _AtomRpc = /*#__PURE__*/_interopRequireWildcard(/*#__PURE__*/require("@effect-atom/atom/AtomRpc"));
exports.AtomRpc = _AtomRpc;
var _Hydration = /*#__PURE__*/_interopRequireWildcard(/*#__PURE__*/require("@effect-atom/atom/Hydration"));
exports.Hydration = _Hydration;
var _Hooks = /*#__PURE__*/require("./Hooks.js");
Object.keys(_Hooks).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _Hooks[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _Hooks[key];
    }
  });
});
var _RegistryContext = /*#__PURE__*/require("./RegistryContext.js");
Object.keys(_RegistryContext).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _RegistryContext[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _RegistryContext[key];
    }
  });
});
function _interopRequireWildcard(e, t) {
  if ("function" == typeof WeakMap) var r = new WeakMap(),
    n = new WeakMap();
  return (_interopRequireWildcard = function (e, t) {
    if (!t && e && e.__esModule) return e;
    var o,
      i,
      f = {
        __proto__: null,
        default: e
      };
    if (null === e || "object" != typeof e && "function" != typeof e) return f;
    if (o = t ? n : r) {
      if (o.has(e)) return o.get(e);
      o.set(e, f);
    }
    for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]);
    return f;
  })(e, t);
}
