// @effect-atom/atom-react/dist/esm/RegistryContext.js
/**
 * @since 1.0.0
 */
"use client";

import * as Registry from "@effect-atom/atom/Registry";
import * as React from "react";
import * as Scheduler from "scheduler";
/**
 * @since 1.0.0
 * @category context
 */
export function scheduleTask(f) {
  Scheduler.unstable_scheduleCallback(Scheduler.unstable_LowPriority, f);
}
/**
 * @since 1.0.0
 * @category context
 */
export const RegistryContext = /*#__PURE__*/React.createContext(/*#__PURE__*/Registry.make({
  scheduleTask,
  defaultIdleTTL: 400
}));
/**
 * @since 1.0.0
 * @category context
 */
export const RegistryProvider = options => {
  const ref = React.useRef(null);
  if (ref.current === null) {
    ref.current = {
      registry: Registry.make({
        scheduleTask: options.scheduleTask ?? scheduleTask,
        initialValues: options.initialValues,
        timeoutResolution: options.timeoutResolution,
        defaultIdleTTL: options.defaultIdleTTL
      })
    };
  }
  React.useEffect(() => {
    if (ref.current?.timeout !== undefined) {
      clearTimeout(ref.current.timeout);
    }
    return () => {
      ref.current.timeout = setTimeout(() => {
        ref.current?.registry.dispose();
        ref.current = null;
      }, 500);
    };
  }, [ref]);
  return React.createElement(RegistryContext.Provider, {
    value: ref.current.registry
  }, options?.children);
};
